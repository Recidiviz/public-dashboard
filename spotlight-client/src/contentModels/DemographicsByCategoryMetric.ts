// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { computed, makeObservable } from "mobx";
import {
  getDemographicCategories,
  recordIsTotalByDimension,
} from "../demographics";
import { DemographicsByCategoryRecord } from "../metricsApi";
import { colors } from "../UiLibrary";
import Metric, { BaseMetricConstructorOptions } from "./Metric";
import { DemographicCategoryRecords } from "./types";

export default class DemographicsByCategoryMetric extends Metric<
  DemographicsByCategoryRecord
> {
  constructor(
    props: BaseMetricConstructorOptions<DemographicsByCategoryRecord>
  ) {
    super(props);

    makeObservable(this, { dataSeries: computed });
  }

  get records(): DemographicsByCategoryRecord[] | undefined {
    let recordsToReturn = this.getOrFetchRecords();
    if (!recordsToReturn) return undefined;

    recordsToReturn = recordsToReturn.filter(
      recordIsTotalByDimension(this.demographicView)
    );
    return recordsToReturn;
  }

  get dataSeries(): DemographicCategoryRecords[] | null {
    const { demographicView, records } = this;
    if (!records || demographicView === "nofilter") return null;

    const categories = getDemographicCategories(demographicView);

    return categories.map(({ identifier, label }) => {
      return {
        label,
        records: records
          .filter((record) =>
            demographicView === "total"
              ? true
              : record[demographicView] === identifier
          )
          .map((record, index) => {
            return {
              label: record.category,
              color: colors.dataViz[index],
              value: record.count,
            };
          }),
      };
    });
  }
}
