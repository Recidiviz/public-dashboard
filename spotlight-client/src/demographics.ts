// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { DemographicFieldKey, DemographicView, TOTAL_KEY } from "./metricsApi";

export const DEMOGRAPHIC_UNKNOWN = "EXTERNAL_UNKNOWN";

const AGE_KEYS = {
  under25: "<25",
  "25_29": "25-29",
  "30_34": "30-34",
  "35_39": "35-39",
  over40: "40<",
};

const AGES = new Map([
  [AGE_KEYS.under25, "<25"],
  [AGE_KEYS["25_29"], "25-29"],
  [AGE_KEYS["30_34"], "30-34"],
  [AGE_KEYS["35_39"], "35-39"],
  [AGE_KEYS.over40, "40<"],
  [DEMOGRAPHIC_UNKNOWN, "Unknown"],
]);

const GENDERS = new Map([
  ["FEMALE", "Female"],
  ["MALE", "Male"],
  [DEMOGRAPHIC_UNKNOWN, "Unknown"],
]);

const RACES = {
  nativeAmerican: "AMERICAN_INDIAN_ALASKAN_NATIVE",
  black: "BLACK",
  hispanic: "HISPANIC",
  white: "WHITE",
  other: "OTHER",
};

// TODO: additional categories that weren't in ND?
const RACE_LABELS = new Map([
  [RACES.nativeAmerican, "Native American"],
  [RACES.black, "Black"],
  [RACES.hispanic, "Hispanic"],
  [RACES.white, "White"],
  [RACES.other, "Other"],
]);

export const DIMENSION_MAPPINGS = new Map<
  Exclude<DemographicView, "nofilter">,
  Map<string, string>
>([
  ["gender", GENDERS],
  ["age", AGES],
  ["race", RACE_LABELS],
  ["total", new Map([[TOTAL_KEY, "Total"]])],
]);

export const DIMENSION_DATA_KEYS: {
  [key in Extract<
    DemographicView,
    "race" | "gender" | "age" | "total"
  >]: DemographicFieldKey;
} = {
  race: "raceOrEthnicity",
  gender: "gender",
  age: "ageBucket",
  // TODO: this shouldn't ever happen but there is type weirdness
  total: "gender",
};
