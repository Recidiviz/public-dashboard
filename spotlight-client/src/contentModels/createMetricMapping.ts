// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { assertNever } from "assert-never";
import { MetricTypeIdList, TenantContent, TenantId } from "../contentApi/types";
import {
  parolePopulationCurrent,
  parolePopulationHistorical,
  paroleProgramParticipationCurrent,
  paroleRevocationReasons,
  paroleSuccessRateDemographics,
  paroleSuccessRateMonthly,
  prisonAdmissionReasons,
  prisonPopulationCurrent,
  prisonPopulationHistorical,
  prisonReleaseTypes,
  prisonStayLengths,
  probationPopulationCurrent,
  probationPopulationHistorical,
  probationProgramParticipationCurrent,
  probationRevocationReasons,
  probationSuccessRateDemographics,
  probationSuccessRateMonthly,
  recidivismRateAllFollowup,
  recidivismRateConventionalFollowup,
  sentencePopulationCurrent,
  sentenceTypesCurrent,
} from "../metricsApi";
import { MetricMapping } from "./types";
import DemographicsByCategoryMetric from "./DemographicsByCategoryMetric";
import HistoricalPopulationBreakdownMetric from "./HistoricalPopulationBreakdownMetric";
import PopulationBreakdownByLocationMetric from "./PopulationBreakdownByLocationMetric";
import ProgramParticipationCurrentMetric from "./ProgramParticipationCurrentMetric";
import RecidivismRateMetric from "./RecidivismRateMetric";
import SentenceTypeByLocationMetric from "./SentenceTypeByLocationMetric";
import SupervisionSuccessRateDemographicsMetric from "./SupervisionSuccessRateDemographicsMetric";
import SupervisionSuccessRateMonthlyMetric from "./SupervisionSuccessRateMonthlyMetric";
import { ERROR_MESSAGES } from "../constants";
import { NOFILTER_KEY, TOTAL_KEY } from "../demographics";
import { colors } from "../UiLibrary";

type MetricMappingFactoryOptions = {
  localityLabelMapping: TenantContent["localities"];
  metadataMapping: TenantContent["metrics"];
  tenantId: TenantId;
};
/**
 * Factory function for converting a mapping of content objects by metric ID
 * to a mapping of Metric instances by metric ID. Creating the entire mapping at once
 * ensures that each ID maps to the proper Metric type without requiring further
 * type guarding on the part of consumers.
 */
export default function createMetricMapping({
  localityLabelMapping,
  metadataMapping,
  tenantId,
}: MetricMappingFactoryOptions): MetricMapping {
  const metricMapping: MetricMapping = new Map();

  // to maintain type safety we iterate through all of the known metrics;
  // iterating through the metadata object's keys widens the type to `string`,
  // which prevents us from guaranteeing exhaustiveness at the type level
  MetricTypeIdList.forEach((metricType) => {
    // not all metrics are required; metadata object is the source of truth
    // for which metrics to include
    const metadata = metadataMapping[metricType];
    if (!metadata) {
      return;
    }

    // messages for errors we may hit at various points below
    const localityContentError = `${ERROR_MESSAGES.missingRequiredContent} (locality labels)`;
    const totalLabelError = `${ERROR_MESSAGES.missingRequiredContent} (total population label)`;

    // this big ol' switch statement ensures that the type ID string union is properly narrowed,
    // allowing for 1:1 correspondence between the ID and the typed Metric instance.
    switch (metricType) {
      case "SentencePopulationCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Sentencing,
              dataTransformer: sentencePopulationCurrent,
              sourceFileName: "sentence_type_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);

        break;
      case "SentenceTypesCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SentenceTypeByLocationMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Sentencing,
            dataTransformer: sentenceTypesCurrent,
            sourceFileName: "sentence_type_by_district_by_demographics",
          })
        );
        break;
      case "PrisonPopulationCurrent":
        if (!localityLabelMapping?.Prison)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Prison,
              dataTransformer: prisonPopulationCurrent,
              sourceFileName:
                "incarceration_population_by_facility_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ProbationPopulationCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Probation,
              dataTransformer: probationPopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ParolePopulationCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Parole,
              dataTransformer: parolePopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "PrisonPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonPopulationHistorical,
            sourceFileName: "incarceration_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationPopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ParolePopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: parolePopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationProgrammingCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ParoleProgrammingCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ProbationSuccessHistorical":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ParoleSuccessHistorical":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ProbationSuccessAggregate":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ParoleSuccessAggregate":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ProbationRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "ParoleRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: paroleRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "PrisonAdmissionReasonsCurrent":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonAdmissionReasons,
            sourceFileName: "incarceration_population_by_admission_reason",
          })
        );
        break;
      case "PrisonReleaseTypeAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonReleaseTypes,
            sourceFileName: "incarceration_releases_by_type_by_period",
          })
        );
        break;
      case "PrisonRecidivismRateHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateAllFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonRecidivismRateSingleFollowupHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            followUpYears: 3,
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateConventionalFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonStayLengthAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonStayLengths,
            sourceFileName: "incarceration_lengths_by_demographics",
            color: colors.dataVizNamed.get("teal"),
          })
        );
        break;
      default:
        assertNever(metricType);
    }
  });

  return metricMapping;
}
