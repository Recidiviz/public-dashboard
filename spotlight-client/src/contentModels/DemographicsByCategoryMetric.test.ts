// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { runInAction, when } from "mobx";
import { DemographicView } from "../demographics";
import { reactImmediately } from "../testUtils";
import createMetricMapping from "./createMetricMapping";
import DemographicsByCategoryMetric from "./DemographicsByCategoryMetric";
import contentFixture from "./__fixtures__/tenant_content_exhaustive";

const testTenantId = "US_ND";
const testMetricId = "PrisonReleaseTypeAggregate";
const testMetadataMapping = {
  [testMetricId]: contentFixture.metrics[testMetricId],
};

function getTestMetric() {
  return createMetricMapping({
    localityLabelMapping: contentFixture.localities,
    metadataMapping: testMetadataMapping,
    tenantId: testTenantId,
  }).get(testMetricId) as DemographicsByCategoryMetric;
}

test("total data", async () => {
  const metric = getTestMetric();

  metric.populateAllRecords();

  await when(() => Boolean(metric.dataSeries));

  reactImmediately(() => {
    expect(metric.dataSeries).toMatchSnapshot();
  });

  expect.hasAssertions();
});

test.each([["raceOrEthnicity"], ["gender"], ["ageBucket"]] as [
  Exclude<DemographicView, "nofilter">
][])("%s data", async (demographicView) => {
  const metric = getTestMetric();

  metric.populateAllRecords();

  await when(() => Boolean(metric.dataSeries));

  runInAction(() => {
    metric.demographicView = demographicView;
  });

  reactImmediately(() => {
    expect(metric.dataSeries).toMatchSnapshot();
  });

  expect.hasAssertions();
});
