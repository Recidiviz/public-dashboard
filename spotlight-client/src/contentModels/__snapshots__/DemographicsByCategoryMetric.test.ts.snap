// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ageBucket data 1`] = `
Array [
  Object {
    "label": "<25",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.3373015873015873,
        "value": 170,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.0992063492063492,
        "value": 50,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.19444444444444445,
        "value": 98,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.057539682539682536,
        "value": 29,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.3115079365079365,
        "value": 157,
      },
    ],
  },
  Object {
    "label": "25-29",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.33542976939203356,
        "value": 160,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.1949685534591195,
        "value": 93,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.09853249475890985,
        "value": 47,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.05870020964360587,
        "value": 28,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.31236897274633124,
        "value": 149,
      },
    ],
  },
  Object {
    "label": "30-39",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.08125331917153479,
        "value": 153,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.25969198088157197,
        "value": 489,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.32501327668613916,
        "value": 612,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.1524163568773234,
        "value": 287,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.1816250663834307,
        "value": 342,
      },
    ],
  },
  Object {
    "label": "40-49",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.46296296296296297,
        "value": 100,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.25925925925925924,
        "value": 56,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.009259259259259259,
        "value": 2,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.26851851851851855,
        "value": 58,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0,
        "value": 0,
      },
    ],
  },
  Object {
    "label": "50-59",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.038910505836575876,
        "value": 10,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.019455252918287938,
        "value": 5,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.7392996108949417,
        "value": 190,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.1245136186770428,
        "value": 32,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.07782101167315175,
        "value": 20,
      },
    ],
  },
  Object {
    "label": "60-69",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.125,
        "value": 6,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0,
        "value": 0,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.08333333333333333,
        "value": 4,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.6666666666666666,
        "value": 32,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.125,
        "value": 6,
      },
    ],
  },
  Object {
    "label": "70+",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.06422018348623854,
        "value": 7,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.009174311926605505,
        "value": 1,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.14678899082568808,
        "value": 16,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.6788990825688074,
        "value": 74,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.10091743119266056,
        "value": 11,
      },
    ],
  },
]
`;

exports[`gender data 1`] = `
Array [
  Object {
    "label": "Male",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.3116474291710388,
        "value": 594,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.09916054564533054,
        "value": 189,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.33630640083945434,
        "value": 641,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.19464847848898217,
        "value": 371,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.05823714585519412,
        "value": 111,
      },
    ],
  },
  Object {
    "label": "Female",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.05804416403785489,
        "value": 92,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.31167192429022084,
        "value": 494,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.3362776025236593,
        "value": 533,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.1949526813880126,
        "value": 309,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.09905362776025237,
        "value": 157,
      },
    ],
  },
]
`;

exports[`raceOrEthnicity data 1`] = `
Array [
  Object {
    "label": "Native American",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.09933142311365807,
        "value": 104,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.3361986628462273,
        "value": 352,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.3113658070678128,
        "value": 326,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.05826170009551098,
        "value": 61,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.19484240687679083,
        "value": 204,
      },
    ],
  },
  Object {
    "label": "Black",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.335609756097561,
        "value": 344,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.1951219512195122,
        "value": 200,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.31121951219512195,
        "value": 319,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.05853658536585366,
        "value": 60,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.09951219512195122,
        "value": 102,
      },
    ],
  },
  Object {
    "label": "Hispanic",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.05913978494623656,
        "value": 22,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.09946236559139784,
        "value": 37,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.33602150537634407,
        "value": 125,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.3118279569892473,
        "value": 116,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.1935483870967742,
        "value": 72,
      },
    ],
  },
  Object {
    "label": "White",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.3360927152317881,
        "value": 203,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.31125827814569534,
        "value": 188,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.057947019867549666,
        "value": 35,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.19536423841059603,
        "value": 118,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.09933774834437085,
        "value": 60,
      },
    ],
  },
  Object {
    "label": "Other",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.09865470852017937,
        "value": 44,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.3116591928251121,
        "value": 139,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.19506726457399104,
        "value": 87,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.336322869955157,
        "value": 150,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.05829596412556054,
        "value": 26,
      },
    ],
  },
]
`;

exports[`total data 1`] = `
Array [
  Object {
    "label": "Total",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Transfer out of system",
        "pct": 0.099112002291607,
        "value": 346,
      },
      Object {
        "color": "#D9A95F",
        "label": "Sentence completion",
        "pct": 0.05814952735605844,
        "value": 203,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Parole",
        "pct": 0.194786594099112,
        "value": 680,
      },
      Object {
        "color": "#4C6290",
        "label": "Probation",
        "pct": 0.3116585505585792,
        "value": 1088,
      },
      Object {
        "color": "#90AEB5",
        "label": "Death",
        "pct": 0.3362933256946434,
        "value": 1174,
      },
    ],
  },
]
`;
